runtime/globroots.c

```diff
@@ -52,15 +60,22 @@ struct skiplist caml_global_roots_old = SKIPLIST_STATIC_INITIALIZER;
 Caml_inline void caml_insert_global_root(struct skiplist * list, value * r)
 {
   caml_plat_lock_blocking(&roots_mutex);
-  caml_skiplist_insert(list, (uintnat) r, 0);
+  caml_skiplist_insert(list, (uintnat) r, ROOT_PRESENT);
   caml_plat_unlock(&roots_mutex);
 }
 
 Caml_inline void caml_delete_global_root(struct skiplist * list, value * r)
 {
-  caml_plat_lock_blocking(&roots_mutex);
-  caml_skiplist_remove(list, (uintnat) r);
-  caml_plat_unlock(&roots_mutex);
+  if (iterating_roots > 0) {
+    uintnat* p = caml_skiplist_find_ptr(list, (uintnat) r);
+    if (p != NULL) {
+      *p = ROOT_DELETED;
```

Comment:
I think it might be worth adding a comment that this write cannot race (if I understand correctly, that is) with a write (from another thread) corresponding to re-use of an existing skiplist cell in `caml_skiplist_insert` because on this line 72, we always hold the lock, so any caller to that function (from this file, where all such calls are protected by the lock) is going to have to wait.

Reply:
Done.